### PR TITLE
Sils/depr

### PR DESCRIPTION
**Pull request by sils1297**

### CODE REVIEW

File: coalib/bearlib/__init__.py [hunk 1]
```diff
@@ -3,3 +3,45 @@
...
+
+     :param depr_args:     A dictionary of settings as keys and their
+                           deprecated names as values.
+    """
+    def _deprecate_decorator(func):
+
+        def wrapping_function(*args, **kwargs):
+            for arg, deprecated_arg in depr_args.items():
+                if deprecated_arg in kwargs and arg not in kwargs:
+                    print("The setting `{}` is deprecated. Please use `{}` "
```
Comment (hypothesist) [inline coalib/bearlib/__init__.py]: Ideally we'd log this to debug, but there's no way I suppose?

File: coalib/bearlib/__init__.py [hunk 2]
```diff
@@ -3,3 +3,45 @@
...
+     Hello world!
+
+     :param depr_args:     A dictionary of settings as keys and their
+                           deprecated names as values.
+    """
+    def _deprecate_decorator(func):
+
+        def wrapping_function(*args, **kwargs):
+            for arg, deprecated_arg in depr_args.items():
+                if deprecated_arg in kwargs and arg not in kwargs:
```
Comment (hypothesist) [inline coalib/bearlib/__init__.py]: What *does* happen when this is violated? Like `param_1` -> `param_2` and `param_0` -> `param_1`

File: coalib/bearlib/__init__.py (hunk 1 above)
Comment (sils1297) [inline coalib/bearlib/__init__.py]: Not yet.

File: coalib/settings/FunctionMetadata.py [hunk 3]
```diff
@@ -91,6 +91,21 @@ def optional_params(self):
...
+        optional, the alias will default to ``None``.
+
+        :param original:  The name of the new setting.
+        :param alias:     The name of the setting that is deprecated.
+        :raises KeyError: If the new setting doesn't exist in the metadata.
+        """
+        self._optional_params[alias] = (
+            self._optional_params[original]
+            if original in self._optional_params
+            else (*self._non_optional_params[original], None))
```
Comment (Asnelchristian) [inline coalib/settings/FunctionMetadata.py]: this can't work

File: coalib/bearlib/__init__.py (hunk 2 above)
Comment (sils1297) [inline coalib/bearlib/__init__.py]: depends on the order the args are given and astonishingly the python version - from python 3.5 we have the kwargs given ordered so if you did `param_0: param_1, param_1: param_2` that could actually work, not sure though

File: coalib/bearlib/__init__.py [hunk 4]
```diff
@@ -3,3 +3,45 @@
...
+     ...     print(new)
+
+     Now we can simply call the bear with the deprecated setting, we'll get a
+     warning - but it still works!
+
+     >>> run(old="Hello world!")
+     The setting `old` is deprecated. Please use `new` instead.
+     Hello world!
+
+     :param depr_args:     A dictionary of settings as keys and their
```
Comment (hypothesist) [inline coalib/bearlib/__init__.py]: longer whitespace between the arg and the definition than required :P

File: coalib/settings/FunctionMetadata.py (hunk 3 above)
Comment (sils1297) [inline coalib/settings/FunctionMetadata.py]: why not, it works with python 3.5 :)

**Comment by hypothesist** [comment]
@sils1297 I feel this (and a lot of other things) could be in coala-utils. We should do the move sometime soon

File: coalib/settings/FunctionMetadata.py [hunk 5]
```diff
@@ -91,6 +91,21 @@ def optional_params(self):
         """
         return self._filter_out_omitted(self._optional_params)
 
+    def add_alias(self, original, alias):
+        """
+        Adds an alias for the original setting. The alias setting will have
+        the same metadata as the original one. If the original setting is not
+        optional, the alias will default to ``None``.
+
+        :param original:  The name of the new setting.
```
Comment (hypothesist) [inline coalib/settings/FunctionMetadata.py]: Why not just name this `new`?

File: coalib/settings/FunctionMetadata.py [hunk 6]
```diff
@@ -91,6 +91,21 @@ def optional_params(self):
...
+        :param alias:     The name of the setting that is deprecated.
```
Comment (hypothesist) [inline coalib/settings/FunctionMetadata.py]: and this should be `original`, right?

File: coalib/settings/FunctionMetadata.py (hunk 5 above)
Comment (sils1297) [inline coalib/settings/FunctionMetadata.py]: well this is the original, the other one is the alias for the original, it's not something new/old as this is not specific to deprecation. I'l change the doc content.

File: coalib/bearlib/__init__.py (hunk 2 above)
Comment (hypothesist) [inline coalib/bearlib/__init__.py]: You should at least add a comment documenting that we can't do this linked deprecation thing

**Comment by hypothesist** [comment]
ack https://github.com/coala-analyzer/coala/pull/2598/commits/fa8fe22562277762fd73ab3761ad1ec33263839a

File: coalib/bearlib/__init__.py (hunk 2 above)
Comment (Asnelchristian) [inline coalib/bearlib/__init__.py]: i think the only fact that ``deprecate_arg`` is in``kwargs`` is enough

File: coalib/settings/FunctionMetadata.py [hunk 7]
```diff
@@ -91,6 +91,21 @@ def optional_params(self):
...
         return self._filter_out_omitted(self._optional_params)
 
+    def add_alias(self, original, alias):
+        """
+        Adds an alias for the original setting. The alias setting will have
+        the same metadata as the original one. If the original setting is not
+        optional, the alias will default to ``None``.
+
+        :param original:  The name of the original setting.
+        :param alias:     The name of the alias for the original.
```
Comment (Asnelchristian) [inline coalib/settings/FunctionMetadata.py]: does *the old setting name* stand for this?
if so i don't think it is clear enough :sweat_smile:

**Comment by Asnelchristian** [comment]
ack https://github.com/coala-analyzer/coala/pull/2598/commits/fd4b8ef8fe9f2bc2fdfb8dd09936aa1994b84404

**Comment by Asnelchristian** [comment]
@rultor merge

**Pull request merged**